pkg/csi/service/wcpguest/controller.go

```diff
@@ -1281,16 +1295,186 @@ func (c *controller) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshot
+					c.supervisorNamespace, supervisorPVCName)
+			}
+			return nil, err
+		}
+		// Get supervisor VolumeSnapshotClass.
+		var supervisorVolumeSnapshotClass string
+		for param := range req.Parameters {
+			paramName := strings.ToLower(param)
+			if paramName == common.AttributeSupervisorVolumeSnapshotClass {
+				supervisorVolumeSnapshotClass = req.Parameters[param]
+			}
+		}
+		// Generate the supervisor VolumeSnapshot name
+		// Assuming snapshot prefix is "snapshot-"
+		supervisorVolumeSnapshotName := c.tanzukubernetesClusterUID + "-" + req.Name[9:]
+		log.Infof("Determined VolumeSnapshotClass: %s for the supervisor VolumeSnapshot: %s",
+			supervisorVolumeSnapshotClass, supervisorVolumeSnapshotName)
+		log.Infof("Looking for VolumeSnapshot %s in supervisor namespace: %s ..",
+			supervisorVolumeSnapshotName, c.supervisorNamespace)
+
+		_, err = c.supervisorSnapshotterClient.SnapshotV1().VolumeSnapshots(c.supervisorNamespace).Get(
+			ctx, supervisorVolumeSnapshotName, metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				// New createSnapshot request on the guest
+				supVolumeSnapshot := getVolumeSnapshotWithVolumeSnapshotClass(supervisorVolumeSnapshotName,
+					c.supervisorNamespace, supervisorVolumeSnapshotClass, supervisorPVCName)
+				log.Infof("Supervisosr VolumeSnapshot Spec: %+v", supVolumeSnapshot)
+				_, err = c.supervisorSnapshotterClient.SnapshotV1().VolumeSnapshots(
+					c.supervisorNamespace).Create(ctx, supVolumeSnapshot, metav1.CreateOptions{})
+				if err != nil {
+					msg := fmt.Sprintf("failed to create volumesnapshot with name: %s on namespace: %s "+
+						"in supervisorCluster. Error: %+v", supervisorVolumeSnapshotName, c.supervisorNamespace, err)
+					log.Error(msg)
+					return nil, status.Errorf(codes.Internal, msg)
+				}
+				log.Infof("Successfully created VolumeSnapshot %s on the supervisor cluster",
+					supervisorVolumeSnapshotName)
+			} else {
+				msg := fmt.Sprintf("failed to get volumesnapshot with name: %s on "+
+					"namespace: %s in supervisorCluster. Error: %+v",
+					supervisorVolumeSnapshotName, c.supervisorNamespace, err)
+				log.Error(msg)
+				return nil, status.Errorf(codes.Internal, msg)
+			}
+		}
+		// Wait for VolumeSnapshot to be ready to use
+		isReady, err := isVolumeSnapshotInSupervisorClusterReady(ctx, c.supervisorSnapshotterClient,
+			supervisorVolumeSnapshotName, c.supervisorNamespace,
+			time.Duration(getProvisionTimeoutInMin(ctx))*time.Minute)
+		if !isReady {
+			msg := fmt.Sprintf("volumesnapshot: %s on namespace: %s in supervisor cluster was not Ready. "+
+				"Error: %+v", supervisorVolumeSnapshotName, c.supervisorNamespace, err)
+			log.Error(msg)
+			return nil, status.Errorf(codes.Internal, msg)
+		}
+		// The VolumeSnapshot on the supervisor is in a ReadyToUse state.
+		vs, err := c.supervisorSnapshotterClient.SnapshotV1().VolumeSnapshots(c.supervisorNamespace).Get(
```

Comment:
In isVolumeSnapshotInSupervisorClusterReady(), are we already retrieving the latest VolumeSnapshot API object with readyToUse is true?  If so, is it better to return the VolumeSnapshot object there?  Otherwise, we are making another call to API server on line 1372 to retrieve VolumeSnapshot object after just retrieved it?

Reply:
nice catch, fixed it.